funding/manager_test.go

```diff
@@ -3724,3 +3718,88 @@ func testUpfrontFailure(t *testing.T, pkscript []byte, expectErr bool) {
 		require.True(t, ok, "did not receive AcceptChannel")
 	}
 }
+
+// TestFundingManagerFundingDoubleSpend tests that we correctly handles the
+// case where we fail to broadcast the funding transaction, and/or another
+// conflicting transaction gets mined.
+func TestFundingManagerFundingDoubleSpend(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		// publishErr is the error to return from the call to
+		// PublishTransaction.
+		publishErr error
+
+		// fundingTxMined indicates whether we the funding tx or a
+		// conflicting tx should be mined.
+		fundingTxMined bool
+	}{
+		{nil, true},
+		{nil, false},
+		{errors.New("any error"), true},
+		{errors.New("any error"), false},
+		{lnwallet.ErrDoubleSpend, true},
+		{lnwallet.ErrDoubleSpend, false},
+	}
+
+	for _, test := range testCases {
+		t.Run(fmt.Sprintf("publishError=%s-fundingTxMinded=%t", test.publishErr, test.fundingTxMined), func(t *testing.T) {
+			alice, bob := setupFundingManagers(t)
+			defer tearDownFundingManagers(t, alice, bob)
+
+			// Intercept the publish method and return the error set for
+			// this test case.
+			alice.fundingMgr.cfg.PublishTransaction = func(
+				txn *wire.MsgTx, label string) error {
+				alice.publTxChan <- txn
+				return test.publishErr
+			}
+
+			// We will consume the channel updates as we go, so no
+			// buffering is needed.
+			updateChan := make(chan *lnrpc.OpenStatusUpdate)
+
+			// Run through the process of opening the channel, up until the
+			// funding transaction is broadcasted.
+			fundingOutPoint, fundingTx := openChannel(t, alice, bob, 500000, 0, 1,
+				updateChan, true)
+
+			// If the test case is checking the case where the funding tx
+			// successfully gets mined and spends the funding input, notify
+			// that the funding input was spent by the funding tx, if not
+			// that it was spent by a conflicting transaction.
+			var spenderHash *chainhash.Hash
+			var err error
+			if test.fundingTxMined {
+				spenderHash = &fundingOutPoint.Hash
+			} else {
+				spenderHash, err = chainhash.NewHashFromStr("abc")
+				if err != nil {
+					t.Fatalf("unable to create hash: %v", err)
+				}
+			}
+
+			alice.mockNotifier.spendChannel <- &chainntnfs.SpendDetail{
+				SpenderTxHash: spenderHash,
+			}
+
+			// If a transaction that conflicts with the funding tx got
+			// mined, we should receive an update saying the funding flow
+			// was canceled at this point.
+			if !test.fundingTxMined {
+				// Alice should have sent an Error message to Bob.
+				assertErrorSent(t, alice.msgChan)
```

Comment:
In addition we should assert that the channel is no longer seen as pending in the funding state by the funding manager.